File: py/minsk/analysis/syntax/unit.py
from collections.abc import Iterable
from dataclasses import dataclass

from minsk.analysis.syntax.kind import SyntaxKind
from minsk.analysis.syntax.node import SyntaxNode
from minsk.analysis.syntax.statement import StatementSyntax
from minsk.analysis.syntax.token import SyntaxToken


@dataclass(frozen=True)
class CompilationUnitSyntax(SyntaxNode):
    statement: StatementSyntax
    end_of_file_token: SyntaxToken

    @property
    def kind(self) -> SyntaxKind:
        return SyntaxKind.CompilationUnit

    @property
    def children(self) -> Iterable["SyntaxNode"]:
        yield self.statement
        yield self.end_of_file_token

File: rs/minsk/src/analysis/syntax/node/unit.rs
use std::fmt::Display;

use crate::analysis::syntax::kind::SyntaxKind;
use crate::analysis::syntax::token::SyntaxToken;

use super::expression::ExpressionSyntax;
use super::SyntaxNode;

#[derive(Debug)]
pub struct CompilationUnitSyntax {
    pub expression: ExpressionSyntax,
    pub end_of_file_token: SyntaxToken,
}

impl Display for CompilationUnitSyntax {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "CompilationUnit")
    }
}

impl SyntaxNode for CompilationUnitSyntax {
    fn kind(&self) -> SyntaxKind {
        SyntaxKind::CompilationUnit
    }

    fn children(&self) -> Vec<&dyn SyntaxNode> {
        vec![&self.expression, &self.end_of_file_token]
    }
}
impl CompilationUnitSyntax {
    pub fn new(expression: ExpressionSyntax, end_of_file_token: SyntaxToken) -> Self {
        Self {
            expression,
            end_of_file_token,
        }
    }
}
